plugins/providers/virtualbox/cap/configure_disks.rb

```diff
@@ -0,0 +1,262 @@
+          dsk_info = {device: "0", port: "0"}
+
+          disk_images = vm_info.select { |v| v.include?("ImageUUID") && v.include?("SATA Controller") }
+          used_ports = disk_images.keys.map { |k| k.split('-') }.map {|v| v[2].to_i}
+          next_available_port = ((0..(MAX_DISK_NUMBER-1)).to_a - used_ports).first
+
+          dsk_info[:port] = next_available_port.to_s
+          if dsk_info[:port].empty?
+            # This likely only occurs if additional disks have been added outside of Vagrant configuration
+            LOGGER.warn("There are no more available ports to attach disks to for the SATA Controller. Clear up some space on the SATA controller to attach new disks.")
+            raise Vagrant::Errors::VirtualBoxDisksDefinedExceedLimit
+          end
+
+          dsk_info
+        end
+
+        def self.resize_disk(machine, disk_config, defined_disk)
+          machine.ui.detail("Disk '#{disk_config.name}' needs to be resized. Resizing disk...", prefix: true)
+
+          if defined_disk["Storage format"] == "VMDK"
+            LOGGER.warn("Disk type VMDK cannot be resized in VirtualBox. Vagrant will convert disk to VDI format to resize first, and then convert resized disk back to VMDK format")
+
+            # grab disk to be resized port and device number
+            disk_info = machine.provider.driver.get_port_and_device(defined_disk["UUID"])
+            # original disk information in case anything goes wrong during clone/resize
+            original_disk = defined_disk
+            backup_disk_location = "#{original_disk["Location"]}.backup"
+
+            # clone disk to vdi formatted disk
+            vdi_disk_file = machine.provider.driver.vmdk_to_vdi(defined_disk["Location"])
+            # resize vdi
+            machine.provider.driver.resize_disk(vdi_disk_file, disk_config.size.to_i)
+
+            begin
+              # Danger Zone
+
+              # remove and close original volume
+              machine.provider.driver.remove_disk(disk_info[:port], disk_info[:device])
+              # Create a backup of the original disk if something goes wrong
+              LOGGER.warn("Making a backup of the original disk at #{defined_disk["Location"]}")
+              FileUtils.mv(defined_disk["Location"], backup_disk_location)
+
+              # we have to close here, otherwise we can't re-clone after
+              # resizing the vdi disk
+              machine.provider.driver.close_medium(defined_disk["UUID"])
+
+              # clone back to original vmdk format and attach resized disk
+              vmdk_disk_file = machine.provider.driver.vdi_to_vmdk(vdi_disk_file)
+              machine.provider.driver.attach_disk(disk_info[:port], disk_info[:device], vmdk_disk_file, "hdd")
+            rescue Exception => e
+              LOGGER.warn("Vagrant encountered an error while trying to resize a disk. Vagrant will now attempt to reattach and preserve the original disk...")
```

Comment:
This rescue block should be wrapped within a `begin/rescue` with any new exceptions being logged as errors and allowing the original exception to be raised. It might make it a bit cleaner to put the cleanup logic in an isolated method you can call from within here.